fplll/pruner/pruner.h

```diff
@@ -290,47 +390,207 @@ template <class FT> class Pruner
+     In such case, it will try to increase the succ. prob as much as
+     it can and possibly return (1, ..., 1).
+  */
+  void optimize_coefficients_prob_incr(/*io*/ vector<double> &pr);
+
+  /**
+     @brief auxiliary function in optimizing the single enumeraiton time
+     fixing the succ. prob.
+
+     Auxiliary function to optimize the single enumeration time with the
+     constraint that the succ. prob (or expected solutions) is fixed. It
+     is used if the given targeted probaility is smaller than the one
+     computed using pruning parameters. Then one decreases the succ.
+     probability by descreasing the pruning parameters.
+  */
+  void optimize_coefficients_prob_decr(/*io*/ vector<double> &pr);
+
+  /**
+     @brief auxiliary function in optimizing the single enumeraiton time fixing succ. prob.
 
-  /** @brief Compute the cost of a single enumeration */
+     Heuristic tuning procedure which seems to be useful.
+  */
+  void optimize_coefficients_prob_tune(/*io*/ vector<double> &pr);
+
+  /**
+     @brief Main interface to optimize pruning coefficients
+
+     Main interface to optimize pruning coefficients. It will invoke either:
+     optimize_coefficients_cost() or
+     optimize_coefficients_prob()
+     depending on the input "target". If the target is negative (e.g. -1),
+     it calls the first function where the goal is to optimize the
+     single_enum_cost() divided by the succ. probability. If the target
+     is > 0, it calls the second function where the goal is to optimize the
+     single_enum_cost() while fixing the succ. probability == target.
+  */
+  void optimize_coefficients(/*io*/ vector<double> &pr);
 
-  double single_enum_cost(/*i*/ const vector<double> &pr, vector<double> *detailed_cost = nullptr)
+  /**
+      @brief Compute the cost of a single enumeration
+  */
+  double single_enum_cost(/*i*/ const vector<double> &pr, vector<double> *detailed_cost = nullptr,
+                          const bool flag = 0)
   {
-    evec b(n);
+    // cout << "# pr is --> " << pr[0] << endl;
+    evec b(d);
     load_coefficients(b, pr);
-    return single_enum_cost(b, detailed_cost).get_d();
+
+    return single_enum_cost(b, detailed_cost, flag).get_d();
   }
 
   /**
      @brief Cost of repeating enumeration and preprocessing until reaching target
 
-     Compute the cost of r enumeration and (r-1) preprocessing, where r is the required number of
-     retrials to reach target.
+     Compute the cost of r enumeration and (r-1) preprocessing, where r is the
+     required number of retrials to reach target.
+
+     There are two modes depending on the target.
+
+     (1) if target == -1, we will optimize the single_enum_cost/prob.
+
+     (2) if target != -1 (target must also be given),  we will optimize
+     the single_enum_cost while fixing the input target prob. In such case,
+     the repeated_enum_cost() actually returns single_enum_cost() since it
```

Comment:
I think the function should be renamed in this case

Reply:
Yes, this is a hack for the moment. The thing is we used it in the optimization procedure as the target function. But since we now have two types of targets, this function also changes accordingly. Perhaps we could make another function for such task? And only call this function in the end of the optimization to pass the values to PruningParams.

Reply:
Maybe have a private function for calling internally and a public function? I'd like to get rid of the hacks before merging since for me/FPyLLL the class API *is* the public interface.

Reply:
I renamed the previous repeated_enum_cost() as target_function() (since it is what we are optimizing for: this could be a single_enum_cost()). Then there is a new repeated_enum_cost() which returns (single_enum_cost() + preprocessing) / succ. prob (or similar).